### ISSUE TITLE
[BUG] EventHubBufferedProducerClient: high CPU usage when no events are being sent

### ISSUE THREAD

**Issue by andreminelli** [Azure/azure-sdk-for-net#28926]
### Library name and version

Azure.Messaging.EventHubs 5.7.0

### Describe the bug

We have an application with high event producing rate (at least hundreads per second) and after a load test (peak of 25K events in 1 minute) with duration of 3 minutes, we observed that CPU usage keeps almost the same as the event 'processing'.

We compared the very same approach/code/test exchanging to an `EventHubProducerClient` (on same version) and CPU drops seconds after the 'injection' of events finishes. 
In both versions, each queue name has a unique instance of this class/Producer being (re)used to send all events.

### Expected behavior

After the load test has stopped - as the the production of events has stopped too - we expect the CPU usage drops to same levels before the test, after a few seconds.

### Actual behavior

CPU usage keeps at the almost same level after several minutes of the ending of load test. Only after a process restart the CPU usage has got back to normal levels.

### Reproduction Steps

The code is very straighforward (I have edited to remove some internal abstraction details, but nothing relevant):

* Using `EventHubBufferedProducerClient`:

```csharp
    public class SenderQueue<T> : IAsyncDisposable
    {
        private readonly ISerializer<T> _serializer;
        private readonly EventHubBufferedProducerClient _eventHubProducerClient;
        private readonly ILogger _logger;

        public SenderQueue(
            IConfiguration configuration,
            string queueName,
            ISerializer<T> serializer,
            ILogger logger)
        {
            _serializer = serializer;
            _logger = logger;
            _eventHubProducerClient = GetProducerClient(configuration, queueName);
        }

        public virtual async Task EnqueueAsync(T item, CancellationToken cancellationToken = default)
        {
            if (item == null)
                throw new ArgumentNullException(nameof(item));
           var serializedItem = _serializer.Serialize(item);
           await _eventHubProducerClient.EnqueueEventAsync(new EventData(serializedItem), cancellationToken);
        }

        public async ValueTask DisposeAsync()
        {
            _eventHubProducerClient.SendEventBatchSucceededAsync -= SendSuccessfulHandler;
            _eventHubProducerClient.SendEventBatchFailedAsync -= SendFailedHandler;
            await _eventHubProducerClient.DisposeAsync();
        }

        protected EventHubBufferedProducerClient GetProducerClient(IConfiguration configuration, string queueName)
        {
            var eventHubBufferedProducerClient = new EventHubBufferedProducerClient(
                            configuration.AzureEventHubConnectionString,
                            queueName);
            eventHubBufferedProducerClient.SendEventBatchSucceededAsync += SendSuccessfulHandler;
            eventHubBufferedProducerClient.SendEventBatchFailedAsync += SendFailedHandler;

            return eventHubBufferedProducerClient;
        }

        private Task SendSuccessfulHandler(SendEventBatchSucceededEventArgs args)
        {
            _logger.Verbose("Successfully sent {EventBatchCount} events of {ItemType} to EventHub", args.EventBatch.Count, typeof(T).Name);
            return Task.CompletedTask;
        }

        private Task SendFailedHandler(SendEventBatchFailedEventArgs args) 
        {
            _logger.Warning(args.Exception, "Failed to send {EventBatchCount} events of {ItemType} to EventHub", args.EventBatch.Count, typeof(T).Name);
            return Task.CompletedTask;
        }
    }
```

* Using `EventHubProducerClient`:

```csharp
    public class SenderQueue<T> : IAsyncDisposable
    {
        private readonly ISerializer<T> _serializer;
        private readonly EventHubProducerClient _eventHubProducerClient;
        private readonly ILogger _logger;

        public SenderQueue(
            IConfiguration configuration,
            string queueName,
            ISerializer<T> serializer,
            ILogger logger)
        {
            _serializer = serializer;
            _logger = logger;
            _eventHubProducerClient = GetProducerClient(configuration, queueName);
        }

        public virtual async Task EnqueueAsync(T item, CancellationToken cancellationToken = default)
        {
            if (item == null)
                throw new ArgumentNullException(nameof(item));
            var eventArray = BuildEventData(item);
            await _eventHubProducerClient.SendAsync(eventArray, cancellationToken);
        }

        public async ValueTask DisposeAsync() => await _eventHubProducerClient.DisposeAsync();

        protected EventHubProducerClient GetProducerClient(IConfiguration configuration, string queueName)
        {
            return new EventHubProducerClient(
                configuration.AzureEventHubConnectionString,
                queueName);
        }

        private EventData[] BuildEventData(T item)
        {
            var serializedItem = _serializer.Serialize(item);
            return new[] { new EventData(serializedItem) };
        }
    }
```

### Environment

* Windows Server 2016 (OS Version:  10.0.14393)
* .Net Core 3.1 (runtime Microsoft.NETCore.App 3.1.1)
* Visual Studio 2022

**Comment by jsquire** [comment]
Hi @andreminelli.  Thank you for reaching out and we regret that you're experiencing difficulties.   Since the `EventHubBufferedProducerClient` has a deferred publishing model, it will continue working in the background after you've stopped enqueuing events until publishing is complete.  The first thing that I'd like to do is ask if we can confirm that the `TotalBufferedEventCount` is zero and you're not seeing the event handlers called at the time you're seeing the CPU activity.

In the meantime, we'll attempt to repro locally using the scenario that you've described.

> We compared the very same approach/code/test exchanging to an EventHubProducerClient (on same version) and CPU drops seconds after the 'injection' of events finishes.

This is expected, as the `EventHubProducerClient` has a deterministic sending model.   Once `SendAsync` completes, there is no longer activity taking place.  The same is not true of the buffered producer, which will continue to make network calls until all enqueued events are cleared from the buffer.

**Comment by andreminelli** [comment]
>  The first thing that I'd like to do is ask if we can confirm that the TotalBufferedEventCount is zero and you're not seeing the event handlers called at the time you're seeing the CPU activity.

I haven't checked this property - I don't even know it until now - and the success event handler log level wasn't enabled during the tests :(
But EventHub "incoming requests" metric had decrease down to previous levels seconds after the load test has finished.

> This is expected, as the EventHubProducerClient has a deterministic sending model. Once SendAsync completes, there is no longer activity taking place. The same is not true of the buffered producer, which will continue to make network calls until all enqueued events are cleared from the buffer.

I understand that, my point is that, apparently, even after all messages have been sent by `EventHubBufferedProducerClient`, CPU usage hasn't dropped.

I will retest and enable all those 'telemetry' mentioned by you in order to capture more data and allow a better view of the behavior.

**Comment by andreminelli** [comment]
BTW, we don't even need a big amount of events to 'trigger' this high CPU usage. With just a few events the CPU goes to same levels after the load test. The success handler is logging info (for those few events, and only 1 or 2 events are 'batched together' on this logs.

**Comment by mpp-sean-c** [comment]
Have also observed this issue with a very similar pattern to the above.  A small number of events using the EventHubBufferedProducerClient resulted in each of our service instances to climb to circa 100% and remain at that level, despite having no further events to process.

**Comment by jsquire** [comment]
Confirming that we've traced this back to a bug in which the task responsible for inspecting the buffers is not backing off during idle periods due to an error in the condition.   A fix is in progress.

**Comment by jsquire** [comment]
This was fixed by #29276 and will be available in our next release.  _(tenative target, the first week of July)_

**Comment by andreminelli** [comment]
Awesome, thank you!
